### PR TITLE
OGGBundle pipeline: Add basic FileLoader section.

### PR DESCRIPTION
**Pull request by lukasgraf**
This adds a basic `FileLoader` section that loads files into documents based on paths.

The distinction between `opengever.document.document` and `ftw.mail.mail` (for `.eml`) is still to be done, otherwise this should be pretty complete.

@deiferni

### CODE REVIEW

File: opengever/setup/sections/fileloader.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+from opengever.setup.sections.jsonsource import BUNDLE_PATH_KEY
+from opengever.setup.sections.jsonsource import JSON_STATS_KEY
+from zope.annotation.interfaces import IAnnotations
+from zope.interface import classProvides
+from zope.interface import implements
+import logging
+import os.path
+
+
+logger = logging.getLogger('opengever.bsmigration.fileloader')
```
Comment (deiferni) [inline opengever/setup/sections/fileloader.py]: bsmigration 😉

File: opengever/setup/sections/fileloader.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+            pathkey = self.pathkey(*keys)[0]
+
+            if self.key in item:
+                filepath = item[self.key]
+                if filepath is None:
+                    yield item
+                    continue
+
+                if pathkey not in item:
+                    logger.warning("Missing path key for file %s" % filepath)
```
Comment (deiferni) [inline opengever/setup/sections/fileloader.py]: we might want to let a validator section check if the file exists and fail hard here instead

File: opengever/setup/sections/fileloader.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+                    self.stats['errors']['files_not_found'][filepath] = path
+                    yield item
+                    continue
+
+                mimetype, _encoding = guess_type(filepath, strict=False)
+                if mimetype is None:
+                    logger.warning("Unknown mimetype for file %s" % filepath)
+                    mimetype = 'application/octet-stream'
+
+                obj = traverse(self.context, path, None)
```
Comment (deiferni) [inline opengever/setup/sections/fileloader.py]: currently the constructed instance is also available as `item['_object']`

File: opengever/setup/sections/fileloader.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                if obj is None:
+                    logger.warning(
```
Comment (deiferni) [inline opengever/setup/sections/fileloader.py]: this should not happen, ever. maybe also fail hard?

File: opengever/setup/sections/fileloader.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+from zope.interface import classProvides
+from zope.interface import implements
+import logging
+import os.path
+
+
+logger = logging.getLogger('opengever.bsmigration.fileloader')
+
+
+class FileLoaderSection(object):
```
Comment (deiferni) [inline opengever/setup/sections/fileloader.py]: I'd also test the section separately, not only in our pipeline blackbox tests.

File: opengever/setup/sections/fileloader.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        # configurable instead of hard coding it here.
+        self.key = 'filepath'
+        self.pathkey = defaultMatcher(options, 'path-key', name, 'path')
+
+        self.bundle_path = IAnnotations(transmogrifier)[BUNDLE_PATH_KEY]
+
+        # TODO: Handle .eml / ftw.mail.mail
+        self.file_field_class = IDocumentSchema['file']._type
+
+        self.stats = IAnnotations(transmogrifier)[JSON_STATS_KEY]
```
Comment (deiferni) [inline opengever/setup/sections/fileloader.py]: i'd also add a short test-case to the sections tests for correct stats-generation

File: opengever/setup/sections/fileloader.py (hunk 1 above)
Comment (lukasgraf) [inline opengever/setup/sections/fileloader.py]: D'oh 🙈

File: opengever/setup/sections/fileloader.py (hunk 5 above)
Comment (lukasgraf) [inline opengever/setup/sections/fileloader.py]: 👍 Done.

File: opengever/setup/sections/fileloader.py (hunk 6 above)
Comment (lukasgraf) [inline opengever/setup/sections/fileloader.py]: 👍 Done

File: opengever/setup/sections/fileloader.py (hunk 2 above)
Comment (lukasgraf) [inline opengever/setup/sections/fileloader.py]: I agree. Until we actually have implemented this as a stand-alone validator, I'd prefer to leave it here, just so we can keep track of edge cases and validations we still need to do.

File: opengever/setup/sections/fileloader.py (hunk 3 above)
Comment (lukasgraf) [inline opengever/setup/sections/fileloader.py]: Yeah. `traverse()` is a transmogrifier utility function though that traverses to objects without being fooled by acquisition or views.

File: opengever/setup/sections/fileloader.py (hunk 4 above)
Comment (lukasgraf) [inline opengever/setup/sections/fileloader.py]: I agree. In the beginning to I think it's useful to keep track of the things that could happen, and Should Not Ever Happen [tm], and then later ramp up the strictness once we have a pre-validation stage in place, and maybe also introduce a strict mode.

**Comment by lukasgraf** [comment]
@deiferni / @phgross updated

**Comment by lukasgraf** [comment]
@deiferni replaced `traverse()` with `item['_object']` as discussed

**Pull request merged**